pkg/events/events.go

```diff
@@ -0,0 +1,66 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package events
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/flowcontrol"
+)
+
+// PodNominationRateLimiter is a pointer so it rate-limits across events
+var PodNominationRateLimiter = flowcontrol.NewTokenBucketRateLimiter(5, 10)
+
+func NominatePod(pod *v1.Pod, node *v1.Node) Event {
+	return Event{
+		InvolvedObject: pod,
+		Type:           v1.EventTypeNormal,
+		Reason:         "Nominate",
+		Message:        fmt.Sprintf("Pod should schedule on %s", node.Name),
+		DedupeValues:   []string{string(pod.UID), node.Name},
+		RateLimiter:    PodNominationRateLimiter,
+	}
+}
+
+func EvictPod(pod *v1.Pod) Event {
+	return Event{
+		InvolvedObject: pod,
+		Type:           v1.EventTypeNormal,
+		Reason:         "Evict",
+		Message:        "Evicted pod",
+		DedupeValues:   []string{pod.Name},
+	}
+}
+
+func PodFailedToSchedule(pod *v1.Pod, err error) Event {
+	return Event{
+		InvolvedObject: pod,
+		Type:           v1.EventTypeWarning,
+		Reason:         "FailedProvisioning",
+		Message:        fmt.Sprintf("Failed to provision new node, %s", err),
+		DedupeValues:   []string{string(pod.UID), err.Error()},
```

Comment:
This is how it was written before. I could conceive that the same error would cause a pod not to schedule, see where the event is being emitted: https://github.com/aws/karpenter-core/blob/main/pkg/controllers/provisioning/scheduling/scheduler.go#L137

Reply:
Ok, seems pretty unlikely to me, but since it was like this before I'll leave as is.